% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/module.overlap.R
\docType{methods}
\name{module.overlap}
\alias{module.overlap}
\title{Plot the overlap among predicted DEG modules}
\usage{
module.overlap(res.module, show.mods = NULL, type = c("model",
  "max.patients", "max.genes")[1], max.n = 30, show.overlap = TRUE,
  cex = 10)
}
\arguments{
\item{res.module}{a 'seed.module' or 'cluster.module' object returned by \code{\link{seed.module}} or \code{\link{cluster.module}}}

\item{show.mods}{the modules to display}

\item{type}{the module type to display}

\item{max.n}{the maximum number of modules to plot. If "show.mods" is set, this option will be ignored.}

\item{show.overlap}{boolen, display the overlap number}

\item{cex}{the font cex to display the overlap number, default 10.}
}
\value{
The heatmap plot for gene and patient overlaps.
}
\description{
Plot the overlap among predicted DEG modules
}
\details{
The DEG modules may have partial overlaps for either genes or patients. This function plots the overlaps for genes and patients from DEG modules in two heatmaps. It can be used to find DEG patterns.
}
\examples{
\dontrun{
module.overlap(res.module, max.n=15,type="model")
}
}
\author{
Guofeng Meng
}
\references{
Gu Z, Eils R and Schlesner M (2016). <U+9225><U+6DD0>omplex heatmaps reveal patterns and correlations in multidimensional genomic data.<U+9225><U+FFFD> Bioinformatics.
}

